dhis2/maps-app
# Open chart as map / open map as chart

This PR allows the user to open a chart in The Data Visualizer as a map, and vice versa. The analytical object is kept in the user data store when the app is switched.

It the analytical object contains multiple data dimensions we open a dialog to allow the user to select between them. One thematic layer will be created per data dimension. 

This PR also adds new relative periods, to keep them synced with Data Visualizer: https://jira.dhis2.org/browse/DHIS2-6066

How to test locally:
- clone the Data visualizer repo: https://github.com/dhis2/data-visualizer-app (master branch)
- yarn install / yarn build / yarn start-app
- create/open a chart 
- open the chart type panel and select "Open as: Map"
- the analytical object should now be stored in the user data store
- open the Maps app with "currentAnalyticalObject": 
http://localhost:8082/?currentAnalyticalObject=true
- test going from Maps app to Data visualizer app by clicking "Open as chart" for a thematic layer. 

Notes:
- Going from a map to a (column) chart is more straightforward than the other way, as this should always work. 
- When we go from a chart to a map we often need to move the dimensions between to rows/columns/filters, and this could result in issues. This should work most of the time :-)
- There are currently limitations when having relative periods as filters, and this could result in strange results - this should be fixed in the backend. 
- If we later add support for multiple period for maps (https://jira.dhis2.org/browse/DHIS2-2255), we could support more chart layouts. 
- Maps app should use the same routing pattern as Data Visualizer (separate issue).
- We should improve error handling across the app (separate issues).

![skjermbilde 2019-03-07 kl 14 37 18](https://user-images.githubusercontent.com/548708/53960432-9bad6a80-40e6-11e9-98c8-1766ca71d4e4.png)

<img width="485" alt="screen shot 2019-03-02 at 15 58 45" src="https://user-images.githubusercontent.com/548708/53683968-a42d2c00-3d07-11e9-9c10-dacc9e4d9a01.png">
